lib/puma/cluster.rb

```diff
@@ -35,34 +35,19 @@ def stop_workers
       @workers.each { |x| x.term }
 
       begin
-        if RUBY_VERSION < '2.6'
-          @workers.each do |w|
+        pids = @workers.map(&:pid)
+        loop do
+          pids.reject! do |w_pid|
             begin
-              Process.waitpid(w.pid)
-            rescue Errno::ECHILD
-              # child is already terminated
-            end
-          end
-        else
-          # below code is for a bug in Ruby 2.6+, above waitpid call hangs
-          t_st = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-          pids = @workers.map(&:pid)
-          loop do
-            pids.reject! do |w_pid|
-              begin
-                if Process.waitpid(w_pid, Process::WNOHANG)
-                  log "    worker status: #{$?}"
-                  true
-                end
-              rescue Errno::ECHILD
-                true # child is already terminated
+              if Process.wait(w_pid, Process::WNOHANG)
+                true
               end
+            rescue Errno::ECHILD
+              true # child is already terminated
             end
-            break if pids.empty?
-            sleep 0.5
           end
-          t_end = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-          log format("    worker shutdown time: %6.2f", t_end - t_st)
+          break if pids.empty?
+          sleep 0.2
```

Comment:
Why do we sleep/wait for everyone to die here but not on line 215?

Reply:
`stop_workers` is just that, it needs to wait for **all the workers to stop**.

`check_workers` is just that, it checks for workers (possibly stopped externally)  that have been `SIGTERM`'d, and removes then from the array, then moves on.  It's not meant to pause, as it essentially runs continuously.

Reply:
One thing I meant to check into is that if a worker is stopped externally, I thought that `check_workers` might possibly wait up to `Const::WORKER_CHECK_INTERVAL` (5 seconds) before it runs `wait` and then respawns?  Is that terrible, or ok?  Or is it running before then by something setting the `force` parameter to true?

Make sense?

A little distracted right now...